File: DSS/ecommerce-analyst/src/commerce_analyst/learning.py
"""Learning weights from measured fix outcomes."""

from __future__ import annotations

from typing import Any, Literal

OutcomeVerdict = Literal["lift", "flat", "worse", "inconclusive"]


def verdict_from_delta(delta_pct: float | None, before_days: int, after_days: int) -> OutcomeVerdict:
    if delta_pct is None or before_days < 2 or after_days < 2:
        return "inconclusive"
    if delta_pct >= 0.05:
        return "lift"
    if delta_pct <= -0.05:
        return "worse"
    return "flat"


def _category_boost(success_rate: float | None, avg_lift: float | None, measured: int) -> float:
    if not measured or success_rate is None:
        return 1.0
    if success_rate >= 0.66 and (avg_lift or 0) > 0.1:
        return 1.12
    if success_rate >= 0.5:
        return 1.06
    if success_rate <= 0.2 and measured >= 2:
        return 0.92
    return 1.0


def compute_learning(actions: list[dict[str, Any]], outcomes: list[dict[str, Any]]) -> dict[str, Any]:
    applied = [action for action in actions if action.get("status") == "applied"]
    outcomes_by_action = {outcome["action_id"]: outcome for outcome in outcomes}
    categories = ["Inventory", "Search", "Funnel", "Anomaly"]
    category_rows = []

    for category in categories:
        related = [action for action in applied if action.get("category") == category]
        measured_outcomes = [
            outcomes_by_action[action["id"]]
            for action in related
            if action["id"] in outcomes_by_action
        ]
        successes = [outcome for outcome in measured_outcomes if outcome.get("verdict") == "lift"]
        lifts = [outcome["delta_pct"] for outcome in measured_outcomes if outcome.get("delta_pct") is not None]
        success_rate = len(successes) / len(measured_outcomes) if measured_outcomes else None
        avg_lift = sum(lifts) / len(lifts) if lifts else None
        boost = _category_boost(success_rate, avg_lift, len(measured_outcomes))
        if measured_outcomes:
            avg_text = f" (avg {avg_lift * 100:.0f}%)" if avg_lift is not None else ""
            insight = (
                f"{category}: {len(successes)}/{len(measured_outcomes)} applied fixes showed lift{avg_text}."
            )
        else:
            insight = f"{category}: no measured outcomes yet — using detector defaults."
        category_rows.append(
            {
                "category": category,
                "attempts": len(related),
                "measured": len(measured_outcomes),
                "successes": len(successes),
                "success_rate": success_rate,
                "avg_lift_pct": avg_lift,
                "boost": boost,
                "insight": insight,
            }
        )

    measured_rows = [row for row in category_rows if row["measured"] > 0]
    global_insight = (
        "Learning loop active across "
        f"{len(measured_rows)} categories. Recommendations are boosted when similar fixes previously lifted target metrics."
        if measured_rows
        else "Learning loop is collecting outcomes. Apply fixes and revisit after a few days to sharpen recommendations."
    )
    return {"categories": category_rows, "global_insight": global_insight}


def apply_learning_boost(opportunities: list[dict[str, Any]], learning: dict[str, Any]) -> list[dict[str, Any]]:
    boosts = {row["category"]: row["boost"] for row in learning["categories"]}
    ranked = []
    for opportunity in opportunities:
        learning_boost = boosts.get(opportunity["category"], 1.0)
        base = opportunity.get("estimated_monthly_revenue_usd")
        if base is None:
            base = opportunity["confidence"] * 1000
        ranked.append(
            {
                **opportunity,
                "learning_boost": learning_boost,
                "adjusted_score": round(base * learning_boost, 2),
            }
        )
    return sorted(ranked, key=lambda item: -(item.get("adjusted_score") or -1))

File: DSS/ecommerce-analyst/dashboard/lib/learning.ts
import type { ActionRecord } from '@/lib/actions';
import type { WeekOpportunity } from '@/lib/week';

export type OutcomeVerdict = 'lift' | 'flat' | 'worse' | 'inconclusive';

export type OutcomeRecord = {
  action_id: string;
  category: string;
  fix_type: string;
  target_metric: string;
  before_value: number | null;
  after_value: number | null;
  delta_pct: number | null;
  verdict: OutcomeVerdict;
  before_days: number;
  after_days: number;
  measured_at: string;
};

export type CategoryLearning = {
  category: string;
  attempts: number;
  measured: number;
  successes: number;
  success_rate: number | null;
  avg_lift_pct: number | null;
  boost: number;
  insight: string;
};

export type LearningSummary = {
  categories: CategoryLearning[];
  global_insight: string;
};

export function verdictFromDelta(deltaPct: number | null, beforeDays: number, afterDays: number): OutcomeVerdict {
  if (deltaPct == null || beforeDays < 2 || afterDays < 2) return 'inconclusive';
  if (deltaPct >= 0.05) return 'lift';
  if (deltaPct <= -0.05) return 'worse';
  return 'flat';
}

export function computeLearning(actions: ActionRecord[], outcomes: OutcomeRecord[]): LearningSummary {
  const applied = actions.filter((action) => action.status === 'applied');
  const outcomesByAction = new Map(outcomes.map((outcome) => [outcome.action_id, outcome]));
  const categories = ['Inventory', 'Search', 'Funnel', 'Anomaly', 'Landing', 'Taxonomy', 'Acquisition'];

  const categoryRows = categories.map((category) => {
    const related = applied.filter((action) => action.category === category);
    const measuredOutcomes = related
      .map((action) => outcomesByAction.get(action.id))
      .filter((outcome): outcome is OutcomeRecord => Boolean(outcome));
    const successes = measuredOutcomes.filter((outcome) => outcome.verdict === 'lift');
    const lifts = measuredOutcomes
      .map((outcome) => outcome.delta_pct)
      .filter((value): value is number => value != null);
    const successRate = measuredOutcomes.length ? successes.length / measuredOutcomes.length : null;
    const avgLift = lifts.length ? lifts.reduce((sum, value) => sum + value, 0) / lifts.length : null;
    const boost = categoryBoost(successRate, avgLift, measuredOutcomes.length);
    const insight = measuredOutcomes.length
      ? `${category}: ${successes.length}/${measuredOutcomes.length} applied fixes showed lift${avgLift != null ? ` (avg ${(avgLift * 100).toFixed(0)}%)` : ''}.`
      : `${category}: no measured outcomes yet — using detector defaults.`;

    return {
      category,
      attempts: related.length,
      measured: measuredOutcomes.length,
      successes: successes.length,
      success_rate: successRate,
      avg_lift_pct: avgLift,
      boost,
      insight,
    };
  });

  const measured = categoryRows.filter((row) => row.measured > 0);
  const globalInsight = measured.length
    ? `Learning loop active across ${measured.length} categories. Recommendations are boosted when similar fixes previously lifted target metrics.`
    : 'Learning loop is collecting outcomes. Apply fixes and revisit after a few days to sharpen recommendations.';

  return { categories: categoryRows, global_insight: globalInsight };
}

function categoryBoost(successRate: number | null, avgLift: number | null, measured: number): number {
  if (!measured || successRate == null) return 1;
  if (successRate >= 0.66 && (avgLift ?? 0) > 0.1) return 1.12;
  if (successRate >= 0.5) return 1.06;
  if (successRate <= 0.2 && measured >= 2) return 0.92;
  return 1;
}

export function applyLearningBoost(opportunities: WeekOpportunity[], learning: LearningSummary): WeekOpportunity[] {
  const boosts = new Map(learning.categories.map((row) => [row.category, row.boost]));
  return opportunities
    .map((opportunity) => {
      const learning_boost = boosts.get(opportunity.category) ?? 1;
      const base = opportunity.estimated_monthly_revenue_usd ?? opportunity.confidence * 1000;
      return {
        ...opportunity,
        learning_boost,
        adjusted_score: Math.round(base * learning_boost * 100) / 100,
      };
    })
    .sort((a, b) => (b.adjusted_score ?? -1) - (a.adjusted_score ?? -1));
}

export function learningContextForPrompt(learning: LearningSummary): string {
  const lines = learning.categories
    .filter((row) => row.measured > 0)
    .map((row) => row.insight);
  return lines.length ? lines.join(' ') : learning.global_insight;
}

export function insightForCategory(learning: LearningSummary, category: string): string | null {
  return learning.categories.find((row) => row.category === category)?.insight ?? null;
}
